### PR TITLE
Add events recorder framework

### LINKED ISSUE #547: Add events recorder framework for recording the events to the resources like errors, warning, info
/kind feature
/area provider/ibmcloud

**Describe the solution you'd like**
Add events to the resources for better information and user can use the `kubectl events` command to view them

**Anything else you would like to add:**
Reference from other provider: https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/main/pkg/record/recorder.go

### PR DESCRIPTION
**Pull request by Prajyot-Parab**
Signed-off-by: Prajyot-Parab <prajyot.parab2@ibm.com>

**What this PR does / why we need it**:
- Add events recorder framework

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #547 

**Special notes for your reviewer**:

/area provider/ibmcloud

1. Please confirm that if this PR changes any image versions, then that's the sole change this PR makes.

**Release note**:

```release-note
Add events recorder framework
```

### CODE REVIEW

File: cloud/scope/cluster.go [hunk 1]
```diff
@@ -102,10 +103,12 @@ func (s *ClusterScope) CreateVPC() (*vpcv1.VPC, error) {
 	options.SetName(s.IBMVPCCluster.Spec.VPC)
 	vpc, _, err := s.IBMVPCClients.VPCService.CreateVPC(options)
 	if err != nil {
+		record.Warnf(s.IBMVPCCluster, "FailedCreate", "Failed vpc creation - %v", err)
 		return nil, err
 	} else if err := s.updateDefaultSG(*vpc.DefaultSecurityGroup.ID); err != nil {
 		return nil, err
```
Comment (mkumatag) [inline cloud/scope/cluster.go:109]: Need Warnf for this failure as well

**Review by mkumatag — CHANGES_REQUESTED** [review]
Overall approach lgmt, need few changes:
- I see the even reason now are more generic like `FailedCreate` or `SuccessfulCreate`, it will be good if we have specific reasons e.g `FailedCreateVPC` or `SuccessfulCreateVPC` for the VPC resources.
- Add fake recorder and add enhance the existing test cases to expect the records.

File: cloud/scope/cluster.go (hunk 1 above)
Comment (Prajyot-Parab) [reply cloud/scope/cluster.go:110]: 1. Added `Warnf` for above failure.
2. Updated record reasons to be more specific.

How about we merge this version of the code and add fake recorder + enhance test cases as part of another PR?

File: cloud/scope/powervs_image.go [hunk 2]
```diff
@@ -163,6 +164,7 @@ func (i *PowerVSImageScope) CreateImageCOSBucket() (*models.ImageReference, *mod
 		return nil, nil, err
```
Comment (mkumatag) [inline cloud/scope/powervs_image.go:164]: do you think adding warning event will be helpful for the failure case here?

File: cloud/scope/powervs_image.go [hunk 3]
```diff
@@ -203,7 +207,12 @@ func (i *PowerVSImageScope) Close() error {
 
 // DeleteImage will delete the image.
 func (i *PowerVSImageScope) DeleteImage() error {
-	return i.IBMPowerVSClient.DeleteImage(i.IBMPowerVSImage.Status.ImageID)
+	if err := i.IBMPowerVSClient.DeleteImage(i.IBMPowerVSImage.Status.ImageID); err != nil {
+		record.Warnf(i.IBMPowerVSImage, "FailedDelete", "Failed image deletion - %v", err)
```
Comment (mkumatag) [inline cloud/scope/powervs_image.go:211]: `FailedDelete?`

File: cloud/scope/powervs_image.go [hunk 4]
```diff
@@ -203,7 +207,12 @@ func (i *PowerVSImageScope) Close() error {
...
+		return err
+	}
+	record.Eventf(i.IBMPowerVSImage, "SuccessfulDelete", "Deleted Image %q", i.IBMPowerVSImage.Status.ImageID)
```
Comment (mkumatag) [inline cloud/scope/powervs_image.go:214]: `SuccessfulDelete?`

File: cloud/scope/powervs_image.go [hunk 5]
```diff
@@ -213,7 +222,12 @@ func (i *PowerVSImageScope) GetImportJob() (*models.Job, error) {
 
 // DeleteImportJob will delete the image import job.
 func (i *PowerVSImageScope) DeleteImportJob() error {
-	return i.IBMPowerVSClient.DeleteJob(i.IBMPowerVSImage.Status.JobID)
+	if err := i.IBMPowerVSClient.DeleteJob(i.IBMPowerVSImage.Status.JobID); err != nil {
+		record.Warnf(i.IBMPowerVSImage, "FailedDelete", "Failed image import job deletion - %v", err)
```
Comment (mkumatag) [inline cloud/scope/powervs_image.go:226]: `FailedDelete?`

File: cloud/scope/powervs_image.go [hunk 6]
```diff
@@ -213,7 +222,12 @@ func (i *PowerVSImageScope) GetImportJob() (*models.Job, error) {
...
+		return err
+	}
+	record.Eventf(i.IBMPowerVSImage, "SuccessfulDelete", "Deleted image import job %q", i.IBMPowerVSImage.Status.JobID)
```
Comment (mkumatag) [inline cloud/scope/powervs_image.go:229]: `SuccessfulDelete?`

File: cloud/scope/cluster.go (hunk 1 above)
Comment (mkumatag) [reply cloud/scope/cluster.go:110]: better to write test cases with the code itself, for this case we can add later

**Review by mkumatag — CHANGES_REQUESTED** [review]
I feel overall there can be more `Success` condition can be added, have that said - this looks good for the initial version. Please fix all the comment and merge this code.

**Review by mkumatag — APPROVED** [review]
/lgtm
/approve

**Pull request merged**